app/src/main/java/com/asfoundation/wallet/transactions/PerkBonusAndGamificationService.kt

```diff
@@ -0,0 +1,298 @@
+import com.appcoins.wallet.gamification.repository.Levels
+import com.appcoins.wallet.gamification.repository.PromotionsRepository
+import com.asf.wallet.R
+import com.asfoundation.wallet.C
+import com.asfoundation.wallet.repository.TransactionRepositoryType
+import com.asfoundation.wallet.ui.TransactionsActivity
+import com.asfoundation.wallet.ui.gamification.GamificationActivity
+import com.asfoundation.wallet.ui.gamification.GamificationMapper
+import com.asfoundation.wallet.ui.gamification.ReachedLevelInfo
+import com.asfoundation.wallet.util.CurrencyFormatUtils
+import com.asfoundation.wallet.util.toBitmap
+import dagger.android.AndroidInjection
+import io.reactivex.Single
+import io.reactivex.functions.Function5
+import io.reactivex.schedulers.Schedulers
+import java.math.BigDecimal
+import java.math.RoundingMode
+import javax.inject.Inject
+import kotlin.math.pow
+
+class PerkBonusAndGamificationService :
+    IntentService(PerkBonusAndGamificationService::class.java.simpleName) {
+
+  @Inject
+  lateinit var transactionRepository: TransactionRepositoryType
+
+  @Inject
+  lateinit var promotionsRepository: PromotionsRepository
+
+  @Inject
+  lateinit var formatter: CurrencyFormatUtils
+
+  @Inject
+  lateinit var gamificationMapper: GamificationMapper
+
+  private lateinit var notificationManager: NotificationManager
+
+
+  override fun onCreate() {
+    super.onCreate()
+    AndroidInjection.inject(this)
+  }
+
+  override fun onHandleIntent(intent: Intent?) {
+    val address = intent?.getStringExtra(ADDRESS_KEY)
+    address?.let {
+      notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+      // this sleep is to wait for all transactions (including perk bonuses) to be registered.
+      // When there are several bonuses associated to the same transaction
+      // it can take a bit of time before every one of them to be inserted in DB
+      Thread.sleep(TRANSACTION_TIME_WAIT_FOR_ALL_IN_MILLIS)
+      // blocking so that service does not call onDestroy beforehand
+      handleNotifications(it).blockingGet()
+    }
+  }
+
+  private fun handleNotifications(address: String): Single<Unit> {
+    return Single.zip(getLastShownLevelUp(address),
+        promotionsRepository.getLastShownLevel(address,
+            GamificationContext.NOTIFICATIONS_ALMOST_NEXT_LEVEL),
+        promotionsRepository.getGamificationStats(address),
+        promotionsRepository.getLevels(address),
+        getNewTransactions(address),
+        Function5 { lastShownLevel: Int, almostNextLevelLastShown: Int, stats: GamificationStats,
+                    allLevels: Levels, transactions: List<Transaction> ->
+          if (isCaseInvalidForNotifications(stats, transactions))
+            return@Function5
+          val maxBonus = allLevels.list.maxBy { level -> level.bonus }!!.bonus
+          val maxLevel = allLevels.list.maxBy { level -> level.level }!!.level
```

Comment:
We should check if levels is empty. It shouldn't if Gamification status is OK, but we never know